### DISCUSSION TITLE
SSL_read returning WANT_READ while poll() returns POLLIN in non-blocking mode

### DISCUSSION THREAD

**Discussion by jay9283743** [openssl/openssl#21924] [Q&A]
What is the best way to handle non-blocking mode when SSL_read returns SSL_ERROR_WANT_READ, but poll() returns data available in the tcp socket ?
Looping SSL_read+poll will cause busy wait since none of them blocks, so I would rather not do that.

Openssl 3.1 (kTLS enabled)

**Comment by mattcaswell** [comment]
If `SSL_read` returns `SSL_ERROR_WANT_READ` and poll says there is data to read then call `SSL_read` again. A failure to return application data from a call to `SSL_read` does not necessarily mean that data has not been consumed from the socket. For example if a NewSessionTicket arrives followed by an app data record, then the first `SSL_read` will read the NewSessionTicket but not produce any app data (and hence return `SSL_ERROR_WANT_READ`). The next call to `SSL_read` will then read the app data record.

**Reply by jay9283743** [reply]
Thanks for your reply.
As far as I can tell, SSL_read permanently fail with SSL_ERROR_WANT_READ while poll permanently return available data/ POLLIN. 
The tcp recv-queue is at 1002 bytes for this particular case

**Reply by jay9283743** [reply]
permanently = Im looping ssl_read + poll

**Reply by mattcaswell** [reply]
That is very odd and should not happen

**Reply by mattcaswell** [reply]
Is this KTLS specific, or does it occur if KTLS is not enabled?

**Reply by jay9283743** [reply]
I'm not sure if it happens without ktls. Strace shows
[pid  3069] recvmsg(260, {msg_namelen=0}, 0) = -1 EAGAIN (Resource temporarily unavailable)
[pid  3069] poll([{fd=260, events=POLLIN}], 1, 300000) = 1 ([{fd=260, revents=POLLIN}])

in loop.

I found recvmsg() in the ktls-code in openssl so it could be relevant to that.
I'm not using recvmsg() in my code

**Reply by mattcaswell** [reply]
`recvmsg` is only either called in the KTLS implementation or in the BIO_s_dgram implementation (which is only relevant for DTLS and I assume you are not using it). So, it does appear to be KTLS related. To confirm it you might want to prevent your application from setting `SSL_OP_ENABLE_KTLS` and see if the problem goes away.

Is this on Linux or FreeBSD?

**Reply by jay9283743** [reply]
Not using dgram, no. Tcp only.
This is linux x64 kernel 5.15 (multiple versions) and 6.1.34.
I disabled ktls on a node now, but in worst case it can take a week+ until it happens

**Comment by asomers** [comment]
I am seeing the same problem on FreeBSD.  OpenSSL 3.1.8 and earlier all worked fine, but the problem appears with 3.2.0 and later.  Just like the OP, `SSL_READ` returns `SSL_ERROR_WANT_READ` forever.  At the OS level, ktrace shows that `poll` returns 1, indicating that the socket is readable.  But then `read` returns 0, indicating EOF.  Also, `netstat` shows that the socket is in the `CLOSE_WAIT` state, meaning that the other side has shut it down.
This seems to me like a bug in openssl.  I would expect `SSL_read` to return `SSL_ERROR_ZERO_RETURN` in this case.
I can readily reproduce the problem if anyone like `mattcaswell` is available to help.
And FTW, my application does not use `SSL_OP_ENABLE_KTLS`.
